core/iwasm/aot/aot_runtime.c

```diff
@@ -2062,94 +2070,103 @@ aot_enlarge_memory(AOTModuleInstance *module_inst, uint32 inc_page_count)
         return false;
     }
 
-    if (total_size >= UINT32_MAX) {
-        return false;
+    bh_assert(total_size_new <= 4 * (uint64)BH_GB);
+    if (total_size_new > UINT32_MAX) {
+        /* Resize to 1 page with size 4G-1 */
+        num_bytes_per_page = UINT32_MAX;
+        total_page_count = max_page_count = 1;
+        total_size_new = UINT32_MAX;
     }
 
 #if WASM_ENABLE_SHARED_MEMORY != 0
-    if (memory_inst->is_shared) {
-        /* For shared memory, we have reserved the maximum spaces during
-            instantiate, only change the cur_page_count here */
-        memory_inst->cur_page_count = total_page_count;
+    if (memory->is_shared) {
+        memory->num_bytes_per_page = UINT32_MAX;
```

Comment:
If the code didn't go in the block between L2075 and L2078, do we still have to change `num_bytes_per_page`?

Reply:
No, only when the memory data size is out of range, or, it reaches 4GB (65536 pages * 65536 bytes_each_page), we need to resize it. As the type of aot_inst->memory_data_size is uint32, and its range is 0 to 4GB-1, and we had better keep the AOT ABI unchanged.